rfcs/rfc-HAARG-0001.md

```diff
@@ -0,0 +1,180 @@
+# Map with different topic variable
+
+## Preamble
+
+    Author:  Graham Knop <haarg@haarg.org>
+    Sponsor:
+    ID: HAARG-0001
+    Status:  Draft
+
+## Abstract
+
+Allow `map` and `grep` to be given a different variable to be used as the topic
+variable, rather than using `$_`. Also allow multiple variables to be given to
+do n-at-a-time iteration.
+
+## Motivation
+
+Traditionally, `map` and `grep` loop over a list, aliasing `$_` to each list
+entry. While this can be convenient in many cases, it gets awkward when nested
+maps are needed. You need to manually save the `$_` value in another variable.
+
+```perl
+my %hash = (
+  foo => [ 1, 2, 3, 4 ],
+  bar => [ 5, 6, 7, 8 ],
+);
+
+my @out = map {
+  my $key = $_;
+  map { "$key: $_" } $hash{$key}->@*;
+} keys %hash;
+```
+
+Using `$_` can also be dangerous if you need to call code not under your
+direct control, due to it being the implicit target of operations like
+`readline`.
+
+It would be more convenient if you could use a different variable for the
+topic, similar to what `for` allows.
+
+```perl
+my @out = map my $key {
+  map my $i { "$key: $i" } $hash{$key}->@*;
+} keys %hash;
+```
+
+A natural extension of this syntax would be to allow n-at-a-time iteration, as
+`for` can do on perl 5.36+.
+
+```perl
+my @out = map my ($key, $val) {
+  map my $i { "$key: $i" } $val->@*;
+} %hash;
+```
+
+All of this would apply to `grep` as well.
+
+As the syntax proposed is currently invalid, a feature should not be needed to
+enable it.
+
+## Rationale
+
+The syntax chosen is meant to follow from the syntax of `for`, treating `map`
+as "`for` as an expression". The chosen syntax does not create any
+ambiguities, and naturally extends to follow the syntax used by `for` for
+n-at-a-time iteration.
+
+## Specification
+
+### `map my VAR BLOCK LIST`
+
+This will evaluate `BLOCK` for each element of `LIST`, aliasing `VAR` to each
+element. Its behavior will otherwise match `map BLOCK LIST`. It is only
+possible to use `my` variables for this.
+
+### `map my (VAR, VAR) BLOCK LIST`
+
+This will evaluate `BLOCK` for each set of two elements in `LIST`, aliasing
+the `VAR` to the first of each set, and the second `VAR` to the second. If
+there is not a multiple of two items in `LIST`, the last run of `BLOCK` will
+assign `undef` to the extra `VAR` slot not corresponding to an element in
+`LIST`. More than two variables can be used to iterate over sets of three or
+more items.
+
```

Comment:
To answer the questions from the comment above, it might be helpful to spell out the behaviour in more detail about `grep my VAR BLOCK LIST` and `grep my (VAR, VAR) BLOCK LIST` here.